In [1]:
import sys
sys.path.append('../../../scripts')
from data_preparation import load_data, convert_datatypes, handle_missing_values, drop_columns, remove_negative_values, calculate_zscore, update_outliers
import pandas as pd

In [2]:
df = load_data('../../../data/benin-malanville.csv')

### Step 1 - Understand the Data

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 525600 entries, 0 to 525599
Data columns (total 19 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Timestamp      525600 non-null  object 
 1   GHI            525600 non-null  float64
 2   DNI            525600 non-null  float64
 3   DHI            525600 non-null  float64
 4   ModA           525600 non-null  float64
 5   ModB           525600 non-null  float64
 6   Tamb           525600 non-null  float64
 7   RH             525600 non-null  float64
 8   WS             525600 non-null  float64
 9   WSgust         525600 non-null  float64
 10  WSstdev        525600 non-null  float64
 11  WD             525600 non-null  float64
 12  WDstdev        525600 non-null  float64
 13  BP             525600 non-null  int64  
 14  Cleaning       525600 non-null  int64  
 15  Precipitation  525600 non-null  float64
 16  TModA          525600 non-null  float64
 17  TModB          525600 non-nul

In [4]:
df.head()

Timestamp  GHI  DNI  DHI  ModA  ModB  Tamb    RH   WS  WSgust  \
0  2021-08-09 00:01 -1.2 -0.2 -1.1   0.0   0.0  26.2  93.4  0.0     0.4   
1  2021-08-09 00:02 -1.1 -0.2 -1.1   0.0   0.0  26.2  93.6  0.0     0.0   
2  2021-08-09 00:03 -1.1 -0.2 -1.1   0.0   0.0  26.2  93.7  0.3     1.1   
3  2021-08-09 00:04 -1.1 -0.1 -1.0   0.0   0.0  26.2  93.3  0.2     0.7   
4  2021-08-09 00:05 -1.0 -0.1 -1.0   0.0   0.0  26.2  93.3  0.1     0.7   

   WSstdev     WD  WDstdev   BP  Cleaning  Precipitation  TModA  TModB  \
0      0.1  122.1      0.0  998         0            0.0   26.3   26.2   
1      0.0    0.0      0.0  998         0            0.0   26.3   26.2   
2      0.5  124.6      1.5  997         0            0.0   26.4   26.2   
3      0.4  120.3      1.3  997         0            0.0   26.4   26.3   
4      0.3  113.2      1.0  997         0            0.0   26.4   26.3   

   Comments  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

In [5]:
df.dtypes

Timestamp         object
GHI              float64
DNI              float64
DHI              float64
ModA             float64
ModB             float64
Tamb             float64
RH               float64
WS               float64
WSgust           float64
WSstdev          float64
WD               float64
WDstdev          float64
BP                 int64
Cleaning           int64
Precipitation    float64
TModA            float64
TModB            float64
Comments         float64
dtype: object

### Step 2 - Clean and Prepare Data 


##### Change datatypes to appropriate formats
In the previous step, the Timestamp colum has a data type of Object. So, here it's changed to date time format.

In [6]:
#convert_datatypes(df)
df['Timestamp'] = pd.to_datetime(df['Timestamp'])
print("Data Type for Timestamp -",df['Timestamp'].dtypes)
df.head()

Data Type for Timestamp - datetime64[ns]


Timestamp  GHI  DNI  DHI  ModA  ModB  Tamb    RH   WS  WSgust  \
0 2021-08-09 00:01:00 -1.2 -0.2 -1.1   0.0   0.0  26.2  93.4  0.0     0.4   
1 2021-08-09 00:02:00 -1.1 -0.2 -1.1   0.0   0.0  26.2  93.6  0.0     0.0   
2 2021-08-09 00:03:00 -1.1 -0.2 -1.1   0.0   0.0  26.2  93.7  0.3     1.1   
3 2021-08-09 00:04:00 -1.1 -0.1 -1.0   0.0   0.0  26.2  93.3  0.2     0.7   
4 2021-08-09 00:05:00 -1.0 -0.1 -1.0   0.0   0.0  26.2  93.3  0.1     0.7   

   WSstdev     WD  WDstdev   BP  Cleaning  Precipitation  TModA  TModB  \
0      0.1  122.1      0.0  998         0            0.0   26.3   26.2   
1      0.0    0.0      0.0  998         0            0.0   26.3   26.2   
2      0.5  124.6      1.5  997         0            0.0   26.4   26.2   
3      0.4  120.3      1.3  997         0            0.0   26.4   26.3   
4      0.3  113.2      1.0  997         0            0.0   26.4   26.3   

   Comments  
0       NaN  
1       NaN  
2       NaN  
3       NaN  
4       NaN

Now, let's check for missing values

In [7]:
#handle_missing_values(df)
print(df.isnull().sum())

Timestamp             0
GHI                   0
DNI                   0
DHI                   0
ModA                  0
ModB                  0
Tamb                  0
RH                    0
WS                    0
WSgust                0
WSstdev               0
WD                    0
WDstdev               0
BP                    0
Cleaning              0
Precipitation         0
TModA                 0
TModB                 0
Comments         525600
dtype: int64


Here we can see that all columns except 'Comments' have non-null values. But the 'Comments' column has null values entirely. So we can go ahead and drop it.

In [8]:
#drop_columns(df)
df = df.drop(['Comments'], axis= 1).copy()
df.shape

(525600, 18)

Now let's look for duplicate values.

In [9]:
print(df.duplicated().sum())

0


We don't have duplicates, so let's move on to handling negative values. The three solar radiation columns, GHI, DNI, DHI have consistent negative values throughout the days. When we take a closer look at the data, there's a direct correlation between these values and night time. And since we're analyzing solar radiation data, we can drop the rows recorded at night or with negative values.

In [10]:
# remove_negative_values(df)
df = df[(df['GHI'] >= 0) & (df['DNI'] >= 0) & (df['DHI'] >= 0)]
df.head(5)

Timestamp   GHI  DNI   DHI  ModA  ModB  Tamb    RH   WS  WSgust  \
413 2021-08-09 06:54:00  16.7  0.0  16.5  16.1  16.3  24.2  98.8  0.0     0.0   
414 2021-08-09 06:55:00  18.2  0.1  18.0  17.4  17.6  24.2  98.8  0.0     0.0   
415 2021-08-09 06:56:00  19.7  0.3  19.5  18.7  18.9  24.2  98.8  0.0     0.0   
416 2021-08-09 06:57:00  21.1  0.6  20.9  19.9  20.1  24.2  98.9  0.0     0.0   
417 2021-08-09 06:58:00  22.5  1.1  22.2  21.1  21.3  24.2  98.9  0.0     0.0   

     WSstdev   WD  WDstdev   BP  Cleaning  Precipitation  TModA  TModB  
413      0.0  0.0      0.0  997         0            0.0   24.2   23.7  
414      0.0  0.0      0.0  997         0            0.0   24.3   23.8  
415      0.0  0.0      0.0  997         0            0.0   24.3   23.9  
416      0.0  0.0      0.0  997         0            0.0   24.4   23.9  
417      0.0  0.0      0.0  997         0            0.0   24.5   24.0

Let's reset the index to start from the 1, now it's aroud 413.

In [11]:
df = df.reset_index(drop=True)
df.head(5)

Timestamp   GHI  DNI   DHI  ModA  ModB  Tamb    RH   WS  WSgust  \
0 2021-08-09 06:54:00  16.7  0.0  16.5  16.1  16.3  24.2  98.8  0.0     0.0   
1 2021-08-09 06:55:00  18.2  0.1  18.0  17.4  17.6  24.2  98.8  0.0     0.0   
2 2021-08-09 06:56:00  19.7  0.3  19.5  18.7  18.9  24.2  98.8  0.0     0.0   
3 2021-08-09 06:57:00  21.1  0.6  20.9  19.9  20.1  24.2  98.9  0.0     0.0   
4 2021-08-09 06:58:00  22.5  1.1  22.2  21.1  21.3  24.2  98.9  0.0     0.0   

   WSstdev   WD  WDstdev   BP  Cleaning  Precipitation  TModA  TModB  
0      0.0  0.0      0.0  997         0            0.0   24.2   23.7  
1      0.0  0.0      0.0  997         0            0.0   24.3   23.8  
2      0.0  0.0      0.0  997         0            0.0   24.3   23.9  
3      0.0  0.0      0.0  997         0            0.0   24.4   23.9  
4      0.0  0.0      0.0  997         0            0.0   24.5   24.0

#### Summary Statistics

In [12]:
df.describe()

Timestamp            GHI            DNI  \
count                         246878  246878.000000  246878.000000   
mean   2022-02-08 16:37:16.496002304     513.411100     356.480524   
min              2021-08-09 06:54:00       0.000000       0.000000   
25%              2021-11-08 14:24:15     239.000000      61.200000   
50%              2022-02-09 13:18:30     520.300000     352.250000   
75%              2022-05-11 15:17:45     790.000000     605.300000   
max              2022-08-08 16:51:00    1413.000000     952.300000   
std                              NaN     304.765569     279.731019   

                 DHI           ModA           ModB           Tamb  \
count  246878.000000  246878.000000  246878.000000  246878.000000   
mean      247.397346     501.775502     485.418619      31.271940   
min         0.000000       0.000000       0.000000      11.000000   
25%       141.200000     218.800000     211.800000      27.700000   
50%       227.800000     502.100000     484.900000      31.500000   
75%       336.200000     786.400000     757.700000      35.400000   
max       759.200000    1342.300000    1342.300000      43.800000   
std       143.420836     307.789277     298.429501       5.459802   

                  RH             WS         WSgust        WSstdev  \
count  246878.000000  246878.000000  246878.000000  246878.000000   
mean       45.541173       2.802063       3.722219       0.587311   
min         2.100000       0.000000       0.000000       0.000000   
25%        18.800000       1.700000       2.400000       0.400000   
50%        44.600000       2.700000       3.600000       0.600000   
75%        68.100000       3.800000       4.900000       0.700000   
max       100.000000      16.000000      22.300000       3.400000   
std        27.015814       1.540762       1.898862       0.249724   

                  WD        WDstdev             BP       Cleaning  \
count  246878.000000  246878.000000  246878.000000  246878.000000   
mean      160.361394      11.926945     994.126139       0.001956   
min         0.000000       0.000000     985.000000       0.000000   
25%        76.500000       8.500000     992.000000       0.000000   
50%       161.300000      11.400000     994.000000       0.000000   
75%       240.900000      14.700000     996.000000       0.000000   
max       360.000000      99.400000    1003.000000       1.000000   
std        95.546187       6.249953       2.637058       0.044188   

       Precipitation          TModA          TModB  
count  246878.000000  246878.000000  246878.000000  
mean        0.000860      47.396643      42.229899  
min         0.000000       9.500000       8.400000  
25%         0.000000      38.800000      34.900000  
50%         0.000000      48.000000      42.400000  
75%         0.000000      56.500000      49.600000  
max         2.500000      81.000000      72.500000  
std         0.030041      12.496610      10.599677

#### Z-Score Analysis

Now, let's move on to outliers. Outliers can be calculated using Z-Score. A Z-Score measures how many standard deviations a data point is from the mean of the dataset. Here we can exclude the Timestamp column because it is consistent and calculating it's Z-Score issues a performance warning due to Adding/subtracting object-dtype array to DatetimeArray not being vectorized.

In [14]:
df_for_zscore = df.drop(['Timestamp'], axis= 1).copy()
df_for_zscore.head(5)

GHI  DNI   DHI  ModA  ModB  Tamb    RH   WS  WSgust  WSstdev   WD  \
0  16.7  0.0  16.5  16.1  16.3  24.2  98.8  0.0     0.0      0.0  0.0   
1  18.2  0.1  18.0  17.4  17.6  24.2  98.8  0.0     0.0      0.0  0.0   
2  19.7  0.3  19.5  18.7  18.9  24.2  98.8  0.0     0.0      0.0  0.0   
3  21.1  0.6  20.9  19.9  20.1  24.2  98.9  0.0     0.0      0.0  0.0   
4  22.5  1.1  22.2  21.1  21.3  24.2  98.9  0.0     0.0      0.0  0.0   

   WDstdev   BP  Cleaning  Precipitation  TModA  TModB  
0      0.0  997         0            0.0   24.2   23.7  
1      0.0  997         0            0.0   24.3   23.8  
2      0.0  997         0            0.0   24.3   23.9  
3      0.0  997         0            0.0   24.4   23.9  
4      0.0  997         0            0.0   24.5   24.0

Since a z-score greater than 3 or less than -3 is considered extreme, here we're filtering outliers greater than the absolute value of 3.

In [ ]:
# calculate_zscore(df)
df_zscores = (df_for_zscore - df_for_zscore.mean()) / df_for_zscore.std()
outliers = df_zscores.abs() > 3
df['Outliers'] = outliers.any(axis=1)
outlier_rows = df[df['Outliers'] == True]
print("Count of rows with outlier values -", df['Outliers'].sum())
print(outlier_rows.head(5))

Count of rows with outlier values - 7713
              Timestamp     GHI    DNI    DHI    ModA    ModB  Tamb    RH  \
220 2021-08-09 10:34:00   872.0  744.5  235.8   801.0   805.0  29.7  71.8   
365 2021-08-09 12:59:00  1088.0  379.7  688.9  1016.0  1023.0  31.7  63.1   
366 2021-08-09 13:00:00  1012.0  298.4  694.3   945.0   951.0  31.4  64.6   
367 2021-08-09 13:01:00  1028.0  308.2  700.2   967.0   973.0  31.2  64.6   
368 2021-08-09 13:02:00   930.0  210.0  702.6   869.0   874.0  31.0  65.8   

      WS  WSgust  WSstdev     WD  WDstdev   BP  Cleaning  Precipitation  \
220  4.0     6.2      1.4  207.7     12.1  999         0            0.0   
365  2.5     3.9      0.6  192.6     26.3  998         0            0.0   
366  5.1     6.4      0.8  211.1      9.6  998         0            0.0   
367  3.8     4.6      0.8  210.4     10.1  998         0            0.0   
368  3.6     4.1      0.5  208.4     14.5  998         0            0.0   

     TModA  TModB  Outliers  
220   64.2   53

Now we can go ahead and replace the outlier values to the column mean for the rows with 'True' value in the 'Outliers' column.

In [16]:
# update_outliers(df)
df_zscore_corrected = df_for_zscore.copy()  
mean = df_for_zscore.mean()

for column in df_for_zscore.columns:
    column_mean = mean[column]
    df_zscore_corrected[column] = df_for_zscore[column].where(~outliers[column], column_mean)

df_zscore_corrected['Timestamp'] = df['Timestamp']
cols = ['Timestamp'] + [col for col in df_zscore_corrected.columns if col != 'Timestamp']
df_zscore_corrected = df_zscore_corrected[cols]

df = df_zscore_corrected
df.head(5)

Timestamp   GHI  DNI   DHI  ModA  ModB  Tamb    RH   WS  WSgust  \
0 2021-08-09 06:54:00  16.7  0.0  16.5  16.1  16.3  24.2  98.8  0.0     0.0   
1 2021-08-09 06:55:00  18.2  0.1  18.0  17.4  17.6  24.2  98.8  0.0     0.0   
2 2021-08-09 06:56:00  19.7  0.3  19.5  18.7  18.9  24.2  98.8  0.0     0.0   
3 2021-08-09 06:57:00  21.1  0.6  20.9  19.9  20.1  24.2  98.9  0.0     0.0   
4 2021-08-09 06:58:00  22.5  1.1  22.2  21.1  21.3  24.2  98.9  0.0     0.0   

   WSstdev   WD  WDstdev     BP  Cleaning  Precipitation  TModA  TModB  
0      0.0  0.0      0.0  997.0       0.0            0.0   24.2   23.7  
1      0.0  0.0      0.0  997.0       0.0            0.0   24.3   23.8  
2      0.0  0.0      0.0  997.0       0.0            0.0   24.3   23.9  
3      0.0  0.0      0.0  997.0       0.0            0.0   24.4   23.9  
4      0.0  0.0      0.0  997.0       0.0            0.0   24.5   24.0

Now let's save our cleaned data.

In [17]:
df.to_csv("../../../data/prepared/benin_prepared.csv")